### PR TITLE
Codegen Non-Native IR Nodes

### LINKED ISSUE #74628: Decouple LTC from TS backend
### 🚀 The feature, motivation and pitch

Currently, there are a few instances where components from the TorchScript LTC backend are used for infrastructure that should be backend agnostic. For example, [`LazyTensor::CreateTensorNode`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/lazy/core/tensor.cpp#L426) relies on `DeviceData`, which is a class based on `TsNode`. This causes some issues when implementing a custom LTC backend, especially when the base class of custom nodes is not `TsNode`.

I'm not sure what the best approach to fix this would be, so I think it would be good to get some other ideas.

CC: @wconstab @antoniojkim 

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by antoniojkim**
Add codegen infrastructure to generate IR nodes for non-native ops.

The proposed change is to add a `non_native` key to the `{backend}_native_functions.yaml` file that contains schema definitions similar to what is found in `native_functions.yaml`. e.g.
```
non_native:
    ...
    - func: expand(Tensor input, int[] size, bool is_scalar_expand) -> Tensor
    ...
```
these definitions are parsed into a `LazyIrSchema` that can be used for generating IR nodes using `GenLazyIR`.

Fixes #74628

CC: @wconstab @desertfire @henrytwo

### CODE REVIEW

File: torchgen/dest/lazy_ir.py [hunk 1]
```diff
@@ -142,29 +157,53 @@ def node_base_ctor_call(self, schema: LazyIrSchema) -> str:
         base_ctor_value_args = ", ".join(base_ctor_value_args_list)
 
         scalar_args = schema.filtered_args(values=False, scalars=True)
-        scalar_hashes = ", ".join([f"{a.name}" for a in scalar_args])
 
-        return f"""{self.node_base}(torch::lazy::OpKind({aten_symbol(schema)}),
-              {{{base_ctor_value_args}}}, std::move(shapes),
+        # Shape constuction.
+        # Conditionally build shape depending on whether it is a native function
+        # and whether the shape cache should be used
```
Comment (wconstab) [inline torchgen/dest/lazy_ir.py:163]: what is the reason to avoid using the shape cache for certain ops?

if it is to align with what is currently happening on master, then I would think we would want to take advantage of doing the codegen and use the shape cache globally.

Comment (antoniojkim) [reply torchgen/dest/lazy_ir.py:163]: Its to make the hash function unique. I was seeing that caching the shapes for some of the non-native IR nodes (and thus not using them to calculate the hash) was causing incorrect cache hits and thus causing errors. It was causing a whole bunch of tests to fail.

Fortunately, the shapes that aren't being cached (marked in `ts_native_functions.yaml`) have trivial shape functions and thus don't really need or benefit from shape cachine

**Comment by antoniojkim** [comment]
> I feel the majority of the ops you are targeting here is view-ops.

This is not exactly true. The purpose of this PR is to provide some codegen mechanism for ops that do not have a corresponding native function. This happens to be all of the view ops. I understand that the functionalization pass is coming soon at which point we won't need to codegen the view ops. However, there are other ops outside of this that benefit from the codegen.

> The point is, if these view-ops will be fully codegen-ed for both backends in the future, I feel this PR is a little overkill for the remaining special ops.

While that is true, as it stands the functionalization pass is not yet ready. So, in the meantime the alternative would be to copy over a 1000 lines of code to another PR which also seems very unnecessary. Having this codegen mechanism means that it can be used in the future for any other additional special ops that are added too.

I think that this helps to clean up a lot of manually written nodes and prevents the need to manually write nodes in the future.

**Comment by antoniojkim** [comment]
@desertfire If the concern about it being "overkill" is enough to stall this PR, I'm fine with only including it in the Torch-MLIR repository and can move it over without much effort. I only wanted to upstream it as it potentially has uses in more than one place. @wconstab thoughts?

**Comment by wconstab** [comment]
I would want to try these simplifications if we merged it,
- to use regular LazyIrSchema/LazyArgument if possible,
- not having such a specialized logic for shape functions

To me it is 'nice to have' but i also don't think it's a big deal to have the duplicated .cpp for now and delete it over time (we don't plan to grow those .cpp much, so it seems maintainable).

**Comment by antoniojkim** [comment]
> to use regular LazyIrSchema/LazyArgument if possible

As far as I'm aware this is not possible as `LazyArgument` only accepts a certain set of C++ types for which there are types in the "non-native" ops that don't belong. That was the reason for introducing `NonNativeLazyArgument` to handle those cases

In order to use `LazyIrSchema` we'd need to make a `FunctionSchema` from the non-native ops. Which is possible but would be far more verbose as the existing parser can't just handle the non-native ops as-is.

The "`NonNative`" versions of these classes subclass the regular classes so there is a lot of shared logic.

> not having such a specialized logic for shape functions

Some of that newly introduced logic is unavoidable as the current `GenLazyIR` class doesn't allow for shapes to be cached. Having said that, I do recognize that a lot of the other stuff does seem very specialized. Though I'm not sure how to remedy this. I'm very much open to any suggestions for how to simplify it.

> To me it is 'nice to have' but i also don't think it's a big deal to have the duplicated .cpp for now and delete it over time (we don't plan to grow those .cpp much, so it seems maintainable).

My main concern with the duplicated .cpp files is that it creates possible points of divergence. _If_ anything does change in PyTorch, this will have to be reflected in Torch-MLIR. However, since the PyTorch team doesn't maintain the Torch-MLIR repository changes can go unnoticed and divergences can occur. Maybe this won't be as big an issue as I'm probably making it out to be, but I'd still rather not have to deal with it in the future.

**Comment by antoniojkim** [comment]
@wconstab Looked a bit more into the simplications that you requested. I was able to create a regular `LazyIrSchema` and `LazyArgument` with some changes to how it's parsed.

Still not sure if there's a better way to deal with the specialized shape logic

**Comment by antoniojkim** [comment]
@wconstab @desertfire I've cleaned up the shape init stuff a bit by introducing an enum. Let me know what you think of the changes

File: torchgen/dest/lazy_ir.py (hunk 1 above)
Comment (desertfire) [reply torchgen/dest/lazy_ir.py:167]: Do you have examples for those failures? I want to understand why they are failing, and the least we can add them as tests.

**Comment by desertfire** [comment]
@JackCaoG , since this also affects pytorch/xla, we need to make sure this does cause any trouble to your migration.

**Comment by JackCaoG** [comment]
PyTorch/XLA will most likely keep our non-native IRs for a while since we expect functionization pass will get rid of most of them and we might need to do something special to `DeviceData` for the SPMD which make it hard to codegen. This change shouldn't block us either, the only concern is it will introduce unintentionally namespace conflicts but it seems like xla test is passing (I assume test got run today since pytorch/xla started to take dependency on codegen today ) .

**Comment by antoniojkim** [comment]
Yeah, so long as you don't specify any `non_native` functions in your native functions yaml, then these changes will basically be a no-op. Though, please feel free to make use of it to clean up and codegen any non-native IRs that can be codegen'd.

>  the only concern is it will introduce unintentionally namespace conflicts but it seems like xla test is passing

I fixed one such namespace issue that came up. The problem might lie going forwards if there are any namespace conflicts that aren't caught because Torch-MLIR isn't part of the PyTorch CI. Hopefully, it won't be too big of a problem. We just need to be careful when modifying any namespaces to change them everywhere.

File: torchgen/dest/lazy_ir.py (hunk 1 above)
Comment (antoniojkim) [reply torchgen/dest/lazy_ir.py:167]: I'm not entirely sure why they specifically are failing. I think likely to do with the fact the shapes are inherent the node's uniqueness. Without which they get "mistaken" for other similar nodes due to false shape cache hits.

For now, I think just keeping the behaviour the same as it was before should be sufficient as changing those does already lead to test failures. Thoughts?

File: torch/csrc/lazy/ts_backend/ts_node.cpp [hunk 2]
```diff
@@ -17,29 +17,29 @@ namespace torch {
...
-  return HashCombine(h, hash_seed);
-}
-
 TsNode::TsNode(OpKind op, OpList operands, std::vector<Shape>&& shapes, size_t num_outputs, hash_t hash_seed)
     : Node(op, operands, std::move(shapes), num_outputs) {
   hash_seed = HashCombine(op.hash(), hash_seed);
-  shape_hash_ = OperandHashes(operands, hash_seed, true);
-  dag_hash_ = (enableDynamicShape() ? OperandHashes(operands, hash_seed, false) : shape_hash_);
+  shape_hash_ = OperandHashes(operands, this->shapes(), hash_seed, true);
+  dag_hash_ = (enableDynamicShape() ? OperandHashes(operands, this->shapes(), hash_seed, false) : shape_hash_);
```
Comment (shunting314) [inline torch/csrc/lazy/ts_backend/ts_node.cpp:42]: The 2 hash handling looks correct to me.

I saw you passed in this->shapes() as well. Looks like we miss the current ops's shape information in the hash before?

Comment (antoniojkim) [reply torch/csrc/lazy/ts_backend/ts_node.cpp:42]: It was included in one of the hash functions, but not the other. So, I just combined to two into a single hash function which makes use of all of the passed in shapes.

Comment (shunting314) [reply torch/csrc/lazy/ts_backend/ts_node.cpp:42]: I think the reason this->shapes() is not passed to OperandHashes is, given OpKind and OpList, the shape of the output(the current) tensor should be deterministic. Thus we don't need to hash it's shape explicitly.

If there are no other reasons, probably we should keep the old behavior. cc @wconstab

Comment (antoniojkim) [reply torch/csrc/lazy/ts_backend/ts_node.cpp:42]: > Thus we don't need to hash it's shape explicitly.

Is your concern that we'll unnecessarily get more cache misses with this? Because I don't think this would affect that. Only ops that don't use the shape cache will have their shape hashes as part of the dag/node hash.

I can kind of see the concern with changing this here as it does affect all of the currently codegen'd native IR nodes. However, once the work is done to make those use the shape cache, I think this will not be a problem.

> given OpKind and OpList, the shape of the output(the current) tensor should be deterministic.

This wasn't the case with some ops that were using this hash function
https://github.com/pytorch/pytorch/blob/6066e5929f12cdef52fe5d85af8001ad3583ec42/torch/csrc/lazy/ts_backend/ts_node.cpp#L33-L36
included the shape hash as well.

Comment (shunting314) [reply torch/csrc/lazy/ts_backend/ts_node.cpp:42]: okay, maybe let's take a step back, why we need change
```
shape_hash_ = OperandHashes(operands,  hash_seed, true);
```
to
```
shape_hash_ = OperandHashes(operands, this->shapes(), hash_seed, true);
```
?

Does the previous implementation cause any problem?  I think your change should work as well. I just want to understand the motivation here.

Comment (wconstab) [reply torch/csrc/lazy/ts_backend/ts_node.cpp:42]: > this->shapes() is not passed to OperandHashes is, given OpKind and OpList, the shape of the output(the current) tensor should be deterministic

Yes, exactly.

This is only untrue for dynamic ops we do not yet support such as 'nonzero'.  For all the currently supported ops, the output shape is a deterministic function of the inputs to the op, for all ops except for the 'leaf' ops.  That's why shape was only hashed on leaf ops and then implicitly propagated thereafter.

I think IIUC antonio's change would add redundant information to the hash.  This means I don't expect any behavior change, but there would be a computational cost associated with the extra hashing.  Anyone think there is actually a behavior change (which I am missing?)

Comment (shunting314) [reply torch/csrc/lazy/ts_backend/ts_node.cpp:42]: I agree with the analysis above.

**Comment by antoniojkim** [comment]
All tests green! @wconstab @desertfire Can I get another review this for PR?

**Comment by desertfire** [comment]
> All tests green! @wconstab @desertfire Can I get another review this for PR?

Can you address Shunting and Will's question about passing shape to GetOpHash?

File: torch/csrc/lazy/ts_backend/ts_node.cpp (hunk 2 above)
Comment (antoniojkim) [reply torch/csrc/lazy/ts_backend/ts_node.cpp:42]: As far as I've seen it doesn't affect the behaviour. The tests appear to confirm that. However, there is that extra hashing cost. Though, if we make the `LazyNativeFunctions` use the shape cache instead of computing the shapes every time, this extra cost disappears.

So, if we're okay with adding the extra cost now with the plan of removing this cost down the line, I'd prefer to keep it as is. However, if there is strong preference to not do this, I can bring back the old hash functions.

The reason I'd prefer to keep it the way it is that I made the non-native IR nodes return whatever their shape inference function returns (i.e. a vector of shapes). So, it will go to the first constructor instead of the leaf constructor. For those ops we need the shapes made as part of the hash. I also think it keeps things cleaner having a single hash function.

@wconstab thoughts?

File: torch/csrc/lazy/backend/backend_interface.h [hunk 3]
```diff
@@ -132,7 +132,7 @@ class TORCH_API BackendRegistrar {
   BackendRegistrar(const BackendImplInterface* backend_impl_interface);
 };
 
-TORCH_API bool hasBackend();
+bool hasBackend();
```
Comment (wconstab) [inline torch/csrc/lazy/backend/backend_interface.h:135]: why is this removed?

File: torchgen/model.py [hunk 4]
```diff
@@ -1026,18 +1026,14 @@ def schema_order_arguments(self) -> Iterator["Argument"]:
             self.arguments.out,
         )
 
+    decl_re = re.compile(r"(?P<name>[^\(]+)\((?P<args>.*)\) -> (?P<returns>.*)")
```
Comment (wconstab) [inline torchgen/model.py:1029]: @bdhirsh are you ok with this?  you should probably take a look at the other codegen changes in this PR and see if you have any issues with it.

File: torchgen/model.py [hunk 5]
```diff
@@ -1328,6 +1324,8 @@ def is_list_like(self) -> Optional["ListType"]:
         "Stream",
         "SymInt",
         "ConstQuantizerPtr",  # TODO: rename
+        "ScalarVal",  # An actual aten Scalar value, no special handling
```
Comment (wconstab) [inline torchgen/model.py:1327]: @antoniojkim what is the reason we had to make a separate 'ScalarVal' when we already had 'Scalar'?

File: torchgen/model.py [hunk 6]
```diff
@@ -1328,6 +1324,8 @@ def is_list_like(self) -> Optional["ListType"]:
...
+        "BackendDataPtr",
```
Comment (wconstab) [inline torchgen/model.py:1328]: @bdhirsh is it ok for us to pollute this list with LT specific types?  it is in service of reusing the parser, which is nice

File: torch/csrc/lazy/backend/backend_interface.h (hunk 3 above)
Comment (antoniojkim) [reply torch/csrc/lazy/backend/backend_interface.h:135]: hmm, not sure. I don't remember removing it. I think it was just probably missed in a rebase. Will revert now

File: torchgen/model.py (hunk 5 above)
Comment (antoniojkim) [reply torchgen/model.py:1327]: `Scalar` is treated specially and is replaced with a `Value` object and handled the same as a `Tensor`. However, for the `Scalar` IR node, we need an actual `c10::Scalar` value that isn't handled specially.

File: torchgen/model.py (hunk 4 above)
Comment (bdhirsh) [reply torchgen/model.py:1029]: The original error messages were a little more specific. But I think this is ok

File: torchgen/model.py (hunk 6 above)
Comment (ezyang) [reply torchgen/model.py:1328]: If you can't put it in an IValue, I don't want it in here. Should I read this PR and find out what it's trying to do more broadly?

File: torchgen/model.py (hunk 5 above)
Comment (antoniojkim) [reply torchgen/model.py:1327]: `foo(Tensor a, Scalar b) -> Tensor` would emit something like
```
foo(Value a, Value b)
```
whereas the intention for `foo(Tensor a, ScalarVal b) -> Tensor` would be to emit something like
```
foo(Value a, Scalar b)
```
This isn't technically just MLIR specific. This is to codegen the `Scalar` IR node which requires specifically to take in a `Scalar` value.

File: torchgen/model.py (hunk 6 above)
Comment (bdhirsh) [reply torchgen/model.py:1328]: Agreed with "can you put it in an IValue" as a good litmus test

File: torchgen/model.py (hunk 5 above)
Comment (bdhirsh) [reply torchgen/model.py:1327]: Got it. One way to handle this without having to reach into and tweak the codegen data model would be with some extra LT-specific metadata for the operator:
```
func: foo(Tensor a, Scalar b) -> Tensor
- treat_scalars_as_constants: True
```

Comment (bdhirsh) [reply torchgen/model.py:1327]: (note we wouldn't want to add this metadata to the `NativeFunction` data class, since that's general codegen. We'd want to plumb this somewhere so it only shows up in LT codegen. Maybe the `LazyIrSchema`?

File: torchgen/model.py (hunk 6 above)
Comment (antoniojkim) [reply torchgen/model.py:1328]: Any thoughts on adding a generic cpp type that isn't parsed but used as is? Something like
```
- func: func_name(::CustomCppType arg_name) -> Tensor
```
which would emit something like
```
func_name(CustomCppType arg_name)
```

Comment (antoniojkim) [reply torchgen/model.py:1331]: I added a proposed solution for this in the last commit. Please check to see if this is an acceptable solution

Comment (bdhirsh) [reply torchgen/model.py:1331]: I'd rather keep a `CustomCppType` out of the data model for the same reason (all of the other codegen, and our JIT type system, wouldn't know how to deal with it, since it's all specific to LTC).

Comment (antoniojkim) [reply torchgen/model.py:1331]: > all of the other codegen, and our JIT type system, wouldn't know how to deal with it

Aren't there checks for edge cases like this? If its not a case thats dealt with I'm generally seeing that there are "unsupported type" errors that are thrown. This should be perfectly valid as those don't support custom C++ types. 

Plus there are no ops outside of the non-native ones that exercise this path. I've designed it to only make a custom C++ type if prefixed with `::` (open to suggestions about how to annotate this. I went with this as it's a recognized namespace pattern). Otherwise, it will fail to parse the type, same as before.

Otherwise, there isn't a way that I can think of at the moment to parse those types using the existing function schema parser. I'd need to re-introduce a non-native op parser which seems overkill now for just a couple unsupported types.

Comment (bdhirsh) [reply torchgen/model.py:1331]: One escape hatch you have is that you can always manually write out the code for the handful of ops that use this new type (how many ops / IR nodes do you plan to use that new data type?)

The codegen today is mainly meant for aten ops. And there is a long term plan of getting general-purpose codegen to work with custom operators, but the idea is that custom operators should work with the rest of the ecosystem. In the same way that today, you can register custom operators to the dispatcher, but there's an expectation that custom ops support conversion to IValues, work with `__torch_dispatch__`, etc.

Comment (ezyang) [reply torchgen/model.py:1331]: We should allow torchbind classes in schema strings. I don't know how to make this work but iirc the schema parsing code does know how to handle custom types, so I'm OK with this for torchgen parser too. (BTW, torchbind passes the IValue test)

Comment (antoniojkim) [reply torchgen/model.py:1328]: I've kept the `DeviceData` as is then as that was the one op that was causing the most issues. Can you please review again to see if there are any other problematic changes?

File: torch/csrc/lazy/ts_backend/ts_node.cpp (hunk 2 above)
Comment (wconstab) [reply torch/csrc/lazy/ts_backend/ts_node.cpp:42]: I think this is OK. If we find we need to optimize it, we have several paths to doing so.

File: torchgen/model.py (hunk 6 above)
Comment (wconstab) [reply torchgen/model.py:1521]: looks ok to me now. thanks for making the change.

Comment (bdhirsh) [reply torchgen/model.py:1521]: Sorry for the late response. Looks like `ScalarVal` and `BackendDataPtr` are still in the enum?

Comment (antoniojkim) [reply torchgen/model.py:1521]: ah, I must have forgotten to remove them after making my changes. Removing now

**Review by wconstab — APPROVED** [review]
Ok, looks good to me @antoniojkim - thanks for your patience and all the updates :)

**Comment by desertfire** [comment]
@antoniojkim , I am not sure if the CI failures relate to your changes. Can you rebase and give it another try? When CI is all green, I will land it.

**Comment by antoniojkim** [comment]
@desertfire The same test seems to be failing. I'm not quite sure what change I made could be causing it though. Any ideas?

**Comment by desertfire** [comment]
> @desertfire The same test seems to be failing. I'm not quite sure what change I made could be causing it though. Any ideas?

I saw your PR contains a change to `third_party/cudnn_frontend`, probably unintentional.

**Comment by antoniojkim** [comment]
> I saw your PR contains a change to `third_party/cudnn_frontend`, probably unintentional.

ah, good catch. Definitely unintentional

**Comment by antoniojkim** [comment]
@desertfire All tests green! Can we get the landing process started?
